cluster/replication/consumer.go

```diff
@@ -451,33 +478,35 @@ func (c *CopyOpConsumer) processFinalizingOp(ctx context.Context, op ShardReplic
 		return api.ShardReplicationState(""), err
 	}
 
-	do := func() bool {
-		asyncReplicationStatus, err := c.replicaCopier.AsyncReplicationStatus(ctx, op.Op.SourceShard.NodeId, op.Op.TargetShard.NodeId, op.Op.SourceShard.CollectionId, op.Op.SourceShard.ShardId)
-		asyncReplicationIsPastUpperTimeBound := asyncReplicationStatus.StartDiffTimeUnixMillis >= asyncReplicationUpperTimeBoundUnixMillis
-		logger.WithFields(logrus.Fields{
-			"err":                                     err,
-			"objects_propagated":                      asyncReplicationStatus.ObjectsPropagated,
-			"start_diff_time_unix_millis":             asyncReplicationStatus.StartDiffTimeUnixMillis,
-			"upper_time_bound_unix_millis":            asyncReplicationUpperTimeBoundUnixMillis,
-			"async_replication_past_upper_time_bound": asyncReplicationIsPastUpperTimeBound,
-		}).Info("async replication finalizing status")
-		return err == nil && asyncReplicationStatus.ObjectsPropagated == 0 && asyncReplicationStatus.StartDiffTimeUnixMillis >= asyncReplicationUpperTimeBoundUnixMillis
-	}
-
 	if ctx.Err() != nil {
 		logger.WithError(ctx.Err()).Debug("error while processing replication operation, shutting down")
 		return api.ShardReplicationState(""), ctx.Err()
 	}
 	// we only check the status of the async replication every 5 seconds to avoid
 	// spamming with too many requests too quickly
+	acceptError := 0
 	err := backoff.Retry(func() error {
-		if ctx.Err() != nil {
-			logger.WithError(ctx.Err()).Debug("error while processing replication operation, shutting down")
-			return backoff.Permanent(ctx.Err())
+		asyncReplicationStatus, err := c.replicaCopier.AsyncReplicationStatus(ctx, op.Op.SourceShard.NodeId, op.Op.TargetShard.NodeId, op.Op.SourceShard.CollectionId, op.Op.SourceShard.ShardId)
```

Comment:
FYI I've created a PR to merge into this branch which DRY's up the "wait for async replication" code and hopefully improves the retry/error handling and logging